src/trace_ng/log.zig

```diff
@@ -0,0 +1,273 @@
+const std = @import("std");
+const Level = @import("level.zig").Level;
+const logfmt = @import("logfmt.zig");
+const sig = @import("../sig.zig");
+// TODO Improve import.
+const Channel = @import("../sync/channel.zig").Channel;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const LogConfig = struct {
+    level: Level = Level.debug,
+    buff_size: usize = 64,
+};
+
+pub const Config = struct {
+    level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    fba_bytes: u64,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const UnscopedLogger = StandardLogger(null);
+pub fn StandardLogger(comptime scope: ?type) type {
+    return struct {
+        const Self = @This();
+        level: Level,
```

Comment:
What do you think about putting the log-level filtering in the interface instead of the implementation? We're probably going to want the same logic for every implementation right? "Do nothing if this log level is turned off."

Here's the tradeoffs I can think of:

pros:
- reduce redundant code
- improved performance since it preempts the need for dynamic dispatch. calling a function pointer is costly, compared to normal function calls

cons:
- unable to put special logic in the implementation that allows you to give some special treatment for excluded logs

Overall I don't think there's a huge difference which way we go.

The performance benefit could be meaningful. Ideally we shouldn't be afraid of improving the debug-level observability to the code. At some point maybe we should add a trace log-level along with scope-filtering for when we want a flood of data for some specific component.

Another idea for performance is comptime log filtering. There could be a be build parameter that results in certain calls to the logger being completely optimized out of the binary. Debug logs would truly have zero cost in that scenario. But this adds another dimension to the configuration which may not be worth the complexity.

The main issue with putting the filter in the interface is that the con is very open-ended. My guess is that it's a non-issue, but maybe in the future we'll suddenly realize that we care about it a lot? I guess we can make changes at that point if we need to.

Reply:
I could not get the interface/vtable approach to work. Ran into issues which I illustrated here https://gist.github.com/dadepo/a0c47dda6fc87aaf737260f1a5961a55 in other not to be blocked, I moved on with the the tagged union approach, which "in theory" should also be performant since it’s a more static type of dispatch.

For most of the re-usable logic, that was moved to the `logfmt` which give specific implementation ability to have custom logic while being able to re-use the ones that are re-usable.